tests/test_pyrflx.py

```diff
@@ -916,30 +936,41 @@ def test_ethernet_generating_ethernet_2_vlan(self) -> None:
             b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
             b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0a"
         )
-        self.frame.set("Destination", int("FFFFFFFFFFFF", 16))
-        self.frame.set("Source", int("0", 16))
-        self.frame.set("Type_Length_TPID", int("8100", 16))
-        self.frame.set("TPID", int("8100", 16))
-        self.frame.set("TCI", 1)
-        self.frame.set("Type_Length", int("0800", 16))
-        self.frame.set("Payload", payload)
-        self.assertTrue(self.frame.valid_message)
+        self.frame_without_refinement.set("Destination", int("FFFFFFFFFFFF", 16))
+        self.frame_without_refinement.set("Source", int("0", 16))
+        self.frame_without_refinement.set("Type_Length_TPID", int("8100", 16))
+        self.frame_without_refinement.set("TPID", int("8100", 16))
+        self.frame_without_refinement.set("TCI", 1)
+        self.frame_without_refinement.set("Type_Length", int("0800", 16))
+        self.frame_without_refinement.set("Payload", payload)
+        self.assertTrue(self.frame_without_refinement.valid_message)
         with open("tests/ethernet_vlan_tag.raw", "rb") as file:
             msg_as_bytes: bytes = file.read()
-        self.assertEqual(self.frame.bytestring, msg_as_bytes)
-
-    # ISSUE: Componolit/RecordFlux#199
+        self.assertEqual(self.frame_without_refinement.bytestring, msg_as_bytes)
 
-    # def test_ethernet_generating_ethernet_2_vlan_dynamic(self) -> None:
-    #    raise NotImplementedError
+    # test is not relevant for the python implementation
+    def test_ethernet_generating_ethernet_2_vlan_dynamic(self) -> None:
+        pass
 
     # rflx in_ipv4-tests.adb
 
-    # def test_ipv4_parsing_udp_in_ipv4(self) -> None:
-    #    raise NotImplementedError
+    def test_ipv4_parsing_udp_in_ipv4(self) -> None:
+        with open("tests/ipv4_udp.raw", "rb") as file:
+            msg_as_bytes: bytes = file.read()
 
-    # def test_ipv4_parsing_udp_in_ipv4_in_ethernet(self) -> None:
-    #    raise NotImplementedError
+        self.ipv4.parse(msg_as_bytes)
+        nested_udp = self.ipv4.get_nested_message("Payload")
+        self.assertTrue(nested_udp.valid_message)
+
+    def test_ipv4_parsing_udp_in_ipv4_in_ethernet(self) -> None:
+        with open("tests/ethernet_ipv4_udp.raw", "rb") as file:
+            msg_as_bytes: bytes = file.read()
+
+        self.frame.parse(msg_as_bytes)
+        nested_ipv4 = self.frame.get_nested_message("Payload")
+        self.assertTrue(nested_ipv4.valid_message)
+        nested_udp = nested_ipv4.get_nested_message("Payload")
+        self.assertTrue(nested_udp.valid_message)
```

Comment:
Is there a way to check which message was contained in Payload? As far as I see I could only check the `name`. In case for IPv4 that would be `Packet`. But if there would be also a refinement to a message `IPv6.Packet`, I couldn't differentiate it from `IPv4.Packet`. I think a `TypeValue` should similarly to `model.Type` contain an identifier, name and package to allow a clear identification.

Reply:
`nested_udp._model.full_name` would return UDP.Datagram. I could introduce a property to TypeValue that holds the model's or type's `full_name` to avoid the protected access. 
Would this suffice for a `full_name`? Would would be the identifier for a TypeValue?

Reply:
Since the introduction of `identifier.ID` identifiers should be preferably used. `full_name` is just string version of an identifier (I also consider removing `full_name`, as it is somehow redundant). I would suggest to add three properties to `TypeValue`: `identifier`, `name` and `package`. All three properties can be get from `_model`.

Reply:
I think I encountered a problem with the identifier in type/model. E.g. UDP.Datagram's (model) Identifier consists of IDs `name` and `parent`. `name` in turn holds another `name` property, a property `parent` and a list `parts`. When evaluating the `parent` property an assertion error "empty identifier" is raised. Also there seems to be a recursion problem as every ID `name` holds another ID `name` and ID `parent`. 
![screenshot](https://user-images.githubusercontent.com/61687542/81394735-b6cc9d80-9122-11ea-9f28-b0ca1d110841.png)

Reply:
I see no issue there. It is intended that `name` and `parent` return again an `ID`. An `ID` consists of "parts". Each part is a string and two parts are separated by a dot.

Examples:

`ID(A.B.C).parent` -> `ID(A.B)`
`ID(A.B.C).name` -> `ID(B)`

`ID(A.B).parent` -> `ID(A)`
`ID(A.B).name` -> `ID(B)`

`ID(A).parent`  -> assertion error as `A` has no parent
`ID(A).name`  -> `ID(A)`

Of course you could do also this:

`ID(A.B).name.name.name.name`  -> `ID(B)`

But why would that be a problem?

Reply:
It just looked like a problem in the debugger, thanks for the explanation.

Reply:
Maybe I found a real problem this time:
While parsing a nested message of type ipv4 i got the following error:
`Error while setting value for field Payload: Error while parsing nested message IPv4.Packet: Error while setting value for field Protocol: 'Number 0 is not a valid enum value'`
In the ipv4 spec, protocol is defined as : 
`type Protocol is (PROTOCOL_UDP => 17) with Size => 8, Always_Valid;`
It occurred while parsing the Ethernet Frame contained in this file: "tests/ethernet_vlan_tag.raw"
It seems to me that "Always_Valid" is not working.

Reply:
You are right, the "Always_Valid" aspect is not handled at all in PyRFLX. If this aspect is set, also unknown values should be accepted during parsing. I suppose the fix should be quite easy. As you seem to need this functionality here, I think it makes sense to fix it in this PR.

Reply:
Added identifier, packet and name in 4a0c236 
Added Always_Valid aspect to EnumValue in e6b322a